src/scvi/external/decipher/_module.py

```diff
@@ -0,0 +1,179 @@
+from collections.abc import Iterable, Sequence
+
+import numpy as np
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+import torch
+import torch.nn as nn
+import torch.utils.data
+from torch.distributions import constraints
+from torch.nn.functional import softmax, softplus
+
+from scvi._constants import REGISTRY_KEYS
+from scvi.module.base import PyroBaseModuleClass, auto_move_data
+
+from ._components import ConditionalDenseNN
+
+
+class DecipherPyroModule(PyroBaseModuleClass):
+    """Pyro Module for the Decipher model.
+
+    This module implements the Decipher model for dimensionality reduction and
+    interpretable representation learning in single-cell RNA sequencing data.
+
+    Parameters
+    ----------
+    dim_genes
+        Number of genes (features) in the dataset.
+    dim_v
+        Dimension of the interpretable latent space v.
+    dim_z
+        Dimension of the intermediate latent space z.
+    layers_v_to_z
+        Hidden layer sizes for the v to z decoder network.
+    layers_z_to_x
+        Hidden layer sizes for the z to x decoder network.
+    beta
+        Regularization parameter for the KL divergence.
+    """
+
+    def __init__(
+        self,
+        dim_genes: int,
+        dim_v: int = 2,
+        dim_z: int = 10,
+        layers_v_to_z: Sequence[int] = (64,),
+        layers_z_to_x: Sequence[int] = (),
+        beta: float = 0.1,
+    ):
+        super().__init__()
+        self.dim_v = dim_v
+        self.dim_z = dim_z
+        self.dim_genes = dim_genes
+        self.layers_v_to_z = layers_v_to_z
+        self.layers_z_to_x = layers_z_to_x
+        self.beta = beta
+
+        self.decoder_v_to_z = ConditionalDenseNN(dim_v, layers_v_to_z, [dim_z] * 2)
+        self.decoder_z_to_x = ConditionalDenseNN(dim_z, layers_z_to_x, [dim_genes])
+        self.encoder_x_to_z = ConditionalDenseNN(dim_genes, [128], [dim_z] * 2)
+        self.encoder_zx_to_v = ConditionalDenseNN(
+            dim_genes + dim_z,
+            [128],
+            [dim_v, dim_v],
+        )
+
+        self.theta = None
+
+        self._epsilon = 1e-5
+
+        self.n_obs = None  # Populated by PyroTrainingPlan
+
+        # Hack: to allow auto_move_data to infer device.
+        self._dummy_param = nn.Parameter(torch.empty(0))
```

Comment:
Can you expand on this?

Reply:
It's likely because you don't have a single parameter?

Reply:
I think it's a relatively clean way to make automove work for pyro modules. The pyro params don't get stored in the same way as normal torch parameters, which is why auto_move_breaks if there are no actual parameters on the class. We do the same thing for the other pyro modules in the repo (e.g., https://github.com/scverse/scvi-tools/blob/main/src/scvi/module/_amortizedlda.py#L95)